Source/x11/XGServerWindow.m

```diff
@@ -4409,75 +4415,140 @@ - (void) freecursor: (void*) cid
+   We map XRandR monitors (outputs) to NSScreen. */
 - (NSArray *)screenList
 {
-  /* I guess screen numbers are in order starting from zero, but we
-     put the main screen first */
- int i;
- int count = ScreenCount(dpy);
- NSMutableArray *screens = [NSMutableArray arrayWithCapacity: count];
- if (count > 0)
-   [screens addObject: [NSNumber numberWithInt: defScreen]];
- for (i = 0; i < count; i++)
-   {
-     if (i != defScreen)
-       [screens addObject: [NSNumber numberWithInt: i]];
-   }
- return screens;
+  NSArray        *screens;
+  NSMutableArray *tmpScreens;
+  int            i;
+
+  monitorsCount = 0;
+  if (monitors != NULL) {
+    NSZoneFree([self zone], monitors);
+  }
+  
+#ifdef HAVE_XRANDR
+  XRRScreenResources  *screen_res;
+  XRROutputInfo       *output_info;
+  XRRCrtcInfo         *crtc_info;
+
+  screen_res = XRRGetScreenResources(dpy, RootWindow(dpy, defScreen));
+  if (screen_res != NULL)
+    {
+      monitorsCount = screen_res->noutput;
+      tmpScreens = [NSMutableArray arrayWithCapacity: monitorsCount];
+      monitors = NSZoneMalloc([self zone], monitorsCount * sizeof(MonitorDevice));
+
+      for (i = 0; i < monitorsCount; i++)
+        {
+          output_info = XRRGetOutputInfo(dpy, screen_res, screen_res->outputs[i]);
+          monitors[i].screen_id = defScreen;
+          if (output_info->crtc)
+            {
+              crtc_info = XRRGetCrtcInfo(dpy, screen_res, output_info->crtc);
+
+              // Fill the cache of device parameters
+              monitors[i].depth = [self windowDepthForScreen: defScreen];
+              monitors[i].resolution = [self resolutionForScreen: defScreen];
+              monitors[i].frame = NSMakeRect(crtc_info->x, crtc_info->y,
+                                             crtc_info->width, crtc_info->height);
+              // Add number of device
+              [tmpScreens addObject: [NSNumber numberWithInt: i]];
+            }
+        }
+      XRRFreeScreenResources(screen_res);
+      screens = [NSArray arrayWithArray: tmpScreens];
+    }
+#endif
+
+  if (monitorsCount == 0)
+    {
+      /* Assumed that it's always available 1 screen per application. We only
+         need to know it's number and it was saved in _initXContext as
+         `defScreen`. */
+      monitorsCount = 1;
+      monitors = NSZoneMalloc([self zone], sizeof(MonitorDevice));
+      monitors[0].screen_id = defScreen;
+      monitors[0].depth = [self windowDepthForScreen: defScreen];
+      monitors[0].resolution = [self resolutionForScreen: defScreen];
+      monitors[0].frame = NSMakeRect(0, 0,
+                                     DisplayWidth(dpy, defScreen),
+                                     DisplayHeight(dpy, defScreen));
+      screens = [NSArray arrayWithObject: [NSNumber numberWithInt: defScreen]];
+    }
+  
+  return screens;
 }
 
-- (NSWindowDepth) windowDepthForScreen: (int) screen_num
+// `screen` is a Xlib screen number.
+- (NSWindowDepth) windowDepthForScreen: (int)screen
```

Comment:
This method will get called by gui as well and there this value will most likely be a monitor number. Please be aware that this method plus a few more are part of the public interface of back that gets used by MS Windows and the upcoming Wayland backend as well. We should not break these methods.

Reply:
You're right. I'll fix this.

Reply:
Done.

Reply:
Now it is even worse. This method gets called when setting up the monitor list. In the XRandR case it gets a screen id, in the old case a monitor id and the list of the monitors is not set up. In your view of the world this method should ignore its parameter and always return the depth of the default screen. And this is in my view wrong as not all monitors have the same depth.

Reply:
The code we're commenting seems outdated. Look here https://github.com/gnustep/libs-back/blob/bcf71f230214873a9621abf295d344a3c8603f69/Source/x11/XGServerWindow.m#L4527. At least my comment in code differs. It states "`screen` is a monitor index not X11 screen". This method should always receive monitor index. As for non-XRandR case monitor list gets set up now (look at bottom of -screenList method). Although I'll fix -windowDepthForScreen call for that case (should send `0` as parameter).

Reply:
Yes, you are right, for some reasons I did see old code here